### PR TITLE
Firebase driver

### PR DESCRIPTION
**Pull request by shans**

### CODE REVIEW

File: src/runtime/storageNG/tests/store-test.ts [hunk 1]
```diff
@@ -57,14 +57,19 @@ describe('Store', async () => {
 
   it(`will throw an exception if an appropriate driver can't be found`, async () => {
     const store = new Store(testKey, Exists.ShouldCreate, null, StorageMode.Direct, CRDTCount);
-    assert.throws(() => store.activate(), 'No driver exists');
+    try {
+      await store.activate();
+      assert.fail('store.activate() should not have succeeded');
```
Comment (csilvestrini) [inline src/runtime/storageNG/tests/store-test.ts]: assertThrowsAsync

File: src/runtime/storageNG/drivers/firebase.ts [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+      this.pendingVersion = currentSnapshot.val().version;
+    }
+      
+    this.reference = reference;
+  }
+
+  registerReceiver(receiver: ReceiveMethod<Data>) {
+    this.receiver = receiver;
+    if (this.pendingModel !== null) {
+      assert(this.pendingModel);
```
[the remark this answers is not in the document]
Comment (shans) [reply src/runtime/storageNG/drivers/firebase.ts]: It confirms it isn't undefined. That could be more explicit I guess.

[the remark this answers is not in the document]
Comment (shans) [reply src/runtime/storageNG/drivers/firebase.ts]: (more context: being undefined should never happen; being null is OK)

File: src/runtime/storageNG/drivers/tests/firebase-test.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    
+    const firebase1 = await FakeFirebaseStorageDriverProvider.newDriverForTesting(firebaseKey, Exists.ShouldCreate);
+    const firebase2 = await FakeFirebaseStorageDriverProvider.newDriverForTesting(firebaseKey, Exists.ShouldExist);
+  });
+
+  it(`can't be instantiated as ShouldExist if the storage location doesn't yet exist`, async () => {
+    const firebaseKey = new FirebaseStorageKey('test-url', 'test-project', 'test-key', 'test-location');
+    try {
+      await FakeFirebaseStorageDriverProvider.newDriverForTesting(firebaseKey, Exists.ShouldExist);
+      assert.fail('should throw exception');
```
[the remark this answers is not in the document]
Comment (shans) [reply src/runtime/storageNG/drivers/tests/firebase-test.ts]: assertThrowsAsync doesn't allow you to check the exception type.

File: src/runtime/storageNG/drivers/tests/firebase-test.ts [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    const firebase1 = await FakeFirebaseStorageDriverProvider.newDriverForTesting<number>(firebaseKey, Exists.ShouldCreate);
+    const recvQueue1: {model: number, version: number}[] = [];
+    firebase1.registerReceiver((model: number, version: number) => recvQueue1.push({model, version}));
+    await firebase1.send(3, 1);
+
+    const firebase2 = await FakeFirebaseStorageDriverProvider.newDriverForTesting<number>(firebaseKey, Exists.ShouldExist);
+    const recvQueue2: {model: number, version: number}[] = [];
+    firebase2.registerReceiver((model: number, version: number) => recvQueue2.push({model, version}));
+
+    assert.equal(recvQueue1.length, 0);
```
[the remark this answers is not in the document]
Comment (shans) [reply src/runtime/storageNG/drivers/tests/firebase-test.ts]: done

File: src/runtime/storageNG/drivers/tests/firebase-test.ts [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+  it(`won't accept out-of-date writes`, async () => {
+    const firebaseKey = new FirebaseStorageKey('test-url', 'test-project', 'test-key', 'test-location');
+    const firebase1 = await FakeFirebaseStorageDriverProvider.newDriverForTesting<number>(firebaseKey, Exists.ShouldCreate);
+
+    assert.equal(true, await firebase1.send(3, 1));
+    assert.equal(false, await firebase1.send(4, 0));
+    assert.equal(false, await firebase1.send(4, 1));
+    assert.equal(false, await firebase1.send(4, 3));
+    assert.equal(true, await firebase1.send(4, 2));
```
[the remark this answers is not in the document]
Comment (shans) [reply src/runtime/storageNG/drivers/tests/firebase-test.ts]: done

File: src/runtime/storageNG/drivers/tests/firebase-test.ts [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    firebase1.registerReceiver((model: number, version: number) => recvQueue1.push({model, version}));
+    const firebase2 = await FakeFirebaseStorageDriverProvider.newDriverForTesting<number>(firebaseKey, Exists.ShouldExist);
+    const recvQueue2: {model: number, version: number}[] = [];
+    firebase2.registerReceiver((model: number, version: number) => recvQueue2.push({model, version}));
+
+    const promise1 = firebase1.send(3, 1);
+    const promise2 = firebase2.send(4, 1);
+
+    const results = await Promise.all([promise1, promise2]);
+    assert.deepEqual([true, false], results);
```
[the remark this answers is not in the document]
Comment (shans) [reply src/runtime/storageNG/drivers/tests/firebase-test.ts]: I can never remember the order with these; I think we get it wrong a *lot* in our code :)

[the remark this answers is not in the document]
Comment (shans) [reply src/runtime/storageNG/drivers/tests/firebase-test.ts]: (done)

**Comment by shans** [comment]
PTAL. I didn't move firebaseKey out because I like the test to read like the steps you'd actually have to do in production code. Not sure about the stop(), maybe it could be moved.

File: src/runtime/storageNG/drivers/tests/firebase-test.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (csilvestrini) [reply src/runtime/storageNG/drivers/tests/firebase-test.ts]: Doesn't it? It includes a `...args` param, and just passes that on to `assert.throws`, which does let you check the exception type / error message:

```
export async function assertThrowsAsync(fn, ...args): Promise<void> {
  try {
    await fn();
    assert.throws(() => undefined, ...args);
  } catch (e) {
    assert.throws(() => {throw e;}, ...args);
  }
}
```

Have you tried doing this: 

```
assertThrowsAsync(
    async () => await FakeFirebaseStorageDriverProvider.newDriverForTesting(firebaseKey, Exists.ShouldExist),
    `Error: requested connection to memory location [object Object] can't proceed as location doesn't exist`);
```

File: src/runtime/storageNG/drivers/firebase.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (csilvestrini) [reply src/runtime/storageNG/drivers/firebase.ts]: Oh, right. I'm still not used to checking for undefined. I'd go with `assert(this.pendingModel !== undefined)` myself, but it's possible this is really obvious to everyone as it is.

**Pull request merged**